### PR TITLE
(WIP)(PUP-6490) Zero sensitive Win32 API strings

### PR DESCRIPTION
**Pull request by Iristyle**
- Needs some tests

/cc @joshcooper

### CODE REVIEW

**Comment by Iristyle** [comment]
This is a quick spike more than anything...

File: lib/puppet/util/windows/api_types.rb [hunk 1]
```diff
@@ -33,6 +33,22 @@ def self.from_string_to_wide_string(str, &block)
       nil
     end
 
+    def self.from_string_to_secure_wide_string(str, &block)
```
Comment (Iristyle) [inline lib/puppet/util/windows/api_types.rb]: I wonder if this is ultimately a little pointless given copies of the same strings floating around in the Ruby managed heap?

**Comment by joshcooper** [comment]
Yeah, I think that's going to be tough.

**Comment by joshcooper** [comment]
Has conflicts and needs to retarget a different branch. I'm also not familiar enough with ruby memory to know if this has the desired effect.

**Comment by McdonaldSeanp** [comment]
needs to be retargeted at 4.10.x, 5.3.x, or master

**Comment by jsane** [comment]
2 cents.. I don't quite follow the ruby syntax but it appears the intent is to zeroize the memory for the wide version of the strings (as generated by from_string_to_secure_wide_strings) when being freed. And with invocations thru FFI ensures that actually happens - that looks reasonable. What am not entirely sure of, as you noted earlier, is whether the original copy of the string still stays around. Perhaps we can try to stomp out their contents as well given that strings in ruby are mutable?

**Comment by joshcooper** [comment]
I'm going to close this for now.

**Pull request closed without merging**